Transform Gizmo Refactor

Dhruv-0-Arora opened the pull request:
## Description
Refactoring the Transform Gizmos to be a `SceneObject`.

### Changes
- All of the transform gizmo computations happen in the `GizmoSceneObject` file. This includes mapping transformations of the mesh to the mirabuf.
- Changed implementation in ejector, intake, and scoring zones configurations.
- Removed implementation from `SceneRenderer`

### Hunter's Additions
- Reworked how a brain is selected for modification/scheme assignment.
- Added new panel specifically for newly spawned in robots.
- Moved scheme selection into the config panel.
- Added `GlobalUIControls` and `GlobalUIComponent` to provide access to context controls outside of react components. Probably not the best practice, but make sense considering these context controls will be relevant for the entire lifetime of the application, not just a single component/menu.
- All current uses of the `GizmoSceneObject` are via the `TransformGizmoControl` component. This allows were easy lifetime controls when using the gizmos.
- Added a move option to the config menu for moving and orientating a robot/field after spawning it.
- Added "Reset Orientation" button to the transform gizmo controls for resetting rotation back to default.
- Changed out default mesh for gizmos to a blank mesh instead of a ball.
- Moved gizmo to the center of the assembly you are moving (center of mass for robots, center of the AA box for fields).
- Moved newly spawned assemblies to `(0, 0 + half of bounding box height, 0)` anchored around the center of their bounding box.
- Removed enter and escape controls for spawning placement/cancellation. Debating adding them back in as the controls for the initial config panel.

[JIRA Issue](https://jira.autodesk.com/browse/AARD-1798)

Dhruv-0-Arora committed:
Adding GizmoSceneObject

Dhruv-0-Arora committed:
Working gizmosceneobject on mirabuf objects
Need to implement elsewhere in the code and also delete old implementation

Dhruv-0-Arora committed:
Fixed implementation of pickup but working on shot trajectory

Dhruv-0-Arora committed:
Fixed Zone Config panel

Dhruv-0-Arora committed:
Removed unused import

Dhruv-0-Arora committed:
Fixed configuring shot trajectory panel

Dhruv-0-Arora committed:
Cleanup

Dhruv-0-Arora committed:
Adding documentation

LucaHaverty reviewed (changes requested):
Couple Issues:

Pressing enter when configuring the intake, ejector, or scoring zones makes the gizmo disappear

Spawning a robot, then deleting the assembly before pressing enter leaves the gizmo behind

Dhruv-0-Arora commented:
> Couple Issues:
> 
> Pressing enter when configuring the intake, ejector, or scoring zones makes the gizmo disappear
> 
> Spawning a robot, then deleting the assembly before pressing enter leaves the gizmo behind

🫠

Dhruv-0-Arora committed:
Added requested changes

BrandonPacewic reviewed (commented):
Looks like this has some pretty extensive merge conflicts from #1068.

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1798/refactoring-gizmos

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1798/refactoring-gizmos

Dhruv-0-Arora committed:
Fixing more merge conflicts

a-crowell reviewed (approved):
One more merge conflict. Other than that, looks good.

BrandonPacewic reviewed (commented):
I'm still not sure that this 100% addresses the concerns that @HunterBarclay had.

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1798/refactoring-gizmos

Dhruv-0-Arora committed:
Renamed mesh to obj and added functional fixes

Dhruv-0-Arora commented:
> Looks like this has some pretty extensive merge conflicts from #1068.

Yep I just added all the refactoring changes that I could remember

Dhruv-0-Arora committed:
Fixed build errors

HunterBarclay commented on fission/src/ui/panels/configuring/assembly-config/interfaces/ConfigureGamepiecePickupInterface.tsx:
```diff
@@ -217,8 +221,8 @@ const ConfigureGamepiecePickupInterface: React.FC<ConfigPickupProps> = ({ select
                         )
-                        transformGizmo.mesh.position.setFromMatrixPosition(robotTransformation)
-                        transformGizmo.mesh.rotation.setFromRotationMatrix(robotTransformation)
+                        transformGizmo.obj.position.setFromMatrixPosition(robotTransformation)
+                        transformGizmo.obj.rotation.setFromRotationMatrix(robotTransformation)
```
Why not build a function inside of GizmoSceneObject to set these? Simply pass in either a matrix or the individual components (translation, rotation, scale) and go from there. Same with getters.

HunterBarclay commented on fission/src/mirabuf/MirabufSceneObject.ts:
```diff
@@ -49,6 +47,8 @@ class MirabufSceneObject extends SceneObject {
     private _ejectable?: EjectableSceneObject
     private _scoringZones: ScoringZoneSceneObject[] = []
 
+    private _gizmo: GizmoSceneObject | undefined
```
Since we've decided that only once gizmo will exist, this should be some kind of static either in the SceneRenderer or GizmoSceneObject itself.

BrandonPacewic replied:
I can think of a case where we would want more than one gizmo in a scene at a time. For instance, what about placing more than one robot during the initialization of match mode?

HunterBarclay replied:
Honestly, I think for initial robot placement, we should move away from using the gizmo tool altogether and move towards Luca's placement for bots for match mode in v6 where you can click and place. In which case, gizmo controls should be reserved for fine movement/orientation/scale control.

HunterBarclay reviewed (changes requested):
In addition to my comment of having the gizmo be handled globally rather than locally with each mirabuf, you should still add the actual react component that can be used to handle the gizmo, and it to the panel for spawning and additional moving.

Take how gizmos are used in zone configuration, make it a react component, use it **everywhere** we use transform gizmos, **including** when spawning in an asset.

Dhruv-0-Arora committed:
Cleaned code
Need to merge to fix caching issue

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1798/refactoring-gizmos

Dhruv-0-Arora committed:
Adding requested changes to remove all translation code from the MirabufSceneObject

Dhruv-0-Arora committed:
Added fix to zoneconfig + added map for all gizmos attached to mirabufSceneObject

PepperLola commented on fission/src/systems/scene/GizmoSceneObject.ts:
```diff
@@ -0,0 +1,194 @@
+        this._gizmo.addEventListener("dragging-changed", (event: { target: TransformControls; value: unknown }) => {
+            // disable orbit controls when dragging the transform gizmo
+            const gizmoDragging = World.SceneRenderer.IsAnyGizmoDragging()
+            if (!event.value && !gizmoDragging) {
+                World.SceneRenderer.orbitControls.enabled = true // enable orbit controls when not dragging another transform gizmo
+            } else if (!event.value && gizmoDragging) {
+                World.SceneRenderer.orbitControls.enabled = false // disable orbit controls when dragging another transform gizmo
+            } else {
+                World.SceneRenderer.orbitControls.enabled = !event.value // disable orbit controls when dragging transform gizmo
+            }
```
```suggestion
            World.SceneRenderer.orbitControls.enabled = !event.value && !gizmoDragging
```

PepperLola commented on fission/src/systems/scene/GizmoSceneObject.ts:
```diff
@@ -0,0 +1,194 @@
+
+            if (event.target.mode === "translate") {
+                // disable other gizmos when translating
+                Array.from(World.SceneRenderer.sceneObjects.values())
+                    .filter(obj => obj instanceof GizmoSceneObject)
+                    .forEach(obj => {
+                        if (obj.gizmo.object === event.target.object && obj.gizmo.mode !== "translate") {
+                            obj.gizmo.dragging = false
+                            obj.gizmo.enabled = !event.value
+                            return
+                        }
+                    })
```
Since you have the `SceneRenderer._gizmosOnMirabuf` map, you should no longer filter `sceneObjects`. In this case, you should make it so that you either make `_gizmosOnMirabuf` publicly available or have a method in SceneRenderer to test if some object has a gizmo (which would be as simple as `this._gizmosOnMirabuf.has(mirabuf_id)` from inside SceneRenderer).

As long as the mirabuf instance id that would be the key in `gizmosOnMirabuf` is available inside this function that I'm commenting on, you can then reduce this from O(n) to O(probably 1) since it would only require a Map lookup instead of an array filter and then foreach.

PepperLola commented on fission/src/systems/scene/GizmoSceneObject.ts:
```diff
@@ -0,0 +1,194 @@
+                event.target.axis = "XYZE"
+            } else if (event.target.mode === "rotate") {
+                // scale on all axes
+                Array.from(World.SceneRenderer.sceneObjects.values())
+                    .filter(obj => obj instanceof GizmoSceneObject)
+                    .forEach(obj => {
+                        if (
+                            obj.gizmo.mode === "scale" &&
+                            event.target !== obj.gizmo &&
+                            obj.gizmo.object === event.target.object
+                        ) {
+                            obj.gizmo.dragging = false
+                            obj.gizmo.enabled = !event.value
+                            return
+                        }
+                    })
```
Same thing here - use the `_gizmosOnMirabuf` map.

PepperLola commented on fission/src/systems/scene/GizmoSceneObject.ts:
```diff
@@ -0,0 +1,194 @@
+        // updating the size of the gizmo based on the distance from the camera
+        const mainCameraFovRadians = (Math.PI * (this._mainCamera.fov * 0.5)) / 180
+        this._gizmo.setSize(
+            (this._size / this._mainCamera.position.distanceTo(this.gizmo.object!.position)) *
```
Are we sure object is always going to be non-null?

PepperLola commented on fission/src/systems/scene/GizmoSceneObject.ts:
```diff
@@ -0,0 +1,194 @@
+        // creating enter key and escape key event listeners
+        if (InputSystem.isKeyPressed("Enter") && this._parentObject) {
+            // confirming placement of object
+            if (this._parentObject !== undefined) this._parentObject.EnablePhysics()
```
```suggestion
            if (this._parentObject) this._parentObject.EnablePhysics()
```

I know `this._parentObject` can only be a `MirabufSceneObject` or `undefined` but I still prefer a truthiness check rather than checking if it's not exactly undefined. It doesn't really matter very much in this instance though.

PepperLola commented on fission/src/systems/scene/GizmoSceneObject.ts:
```diff
@@ -0,0 +1,194 @@
+class GizmoSceneObject extends SceneObject {
+    private _gizmo: TransformControls
+    private _obj: Object3D
+    private _parentObject: MirabufSceneObject | undefined
```
```suggestion
    private _parentObject?: MirabufSceneObject
```

I prefer this just because it's cleaner and is also what you do in the constructor.

PepperLola commented on fission/src/systems/scene/GizmoSceneObject.ts:
```diff
@@ -0,0 +1,194 @@
+                World.SceneRenderer.orbitControls.enabled = !event.value // disable orbit controls when dragging transform gizmo
+            }
+
+            if (event.target.mode === "translate") {
+                // disable other gizmos when translating
+                Array.from(World.SceneRenderer.sceneObjects.values())
+                    .filter(obj => obj instanceof GizmoSceneObject)
+                    .forEach(obj => {
+                        if (obj.gizmo.object === event.target.object && obj.gizmo.mode !== "translate") {
+                            obj.gizmo.dragging = false
+                            obj.gizmo.enabled = !event.value
+                            return
+                        }
+                    })
+            } else if (
+                event.target.mode === "scale" &&
+                (InputSystem.isKeyPressed("ShiftRight") || InputSystem.isKeyPressed("ShiftLeft"))
+            ) {
+                // scale uniformly if shift is pressed
+                event.target.axis = "XYZE"
+            } else if (event.target.mode === "rotate") {
+                // scale on all axes
+                Array.from(World.SceneRenderer.sceneObjects.values())
+                    .filter(obj => obj instanceof GizmoSceneObject)
+                    .forEach(obj => {
+                        if (
+                            obj.gizmo.mode === "scale" &&
+                            event.target !== obj.gizmo &&
+                            obj.gizmo.object === event.target.object
+                        ) {
+                            obj.gizmo.dragging = false
+                            obj.gizmo.enabled = !event.value
+                            return
+                        }
+                    })
+            }
```
Would a switch be better here? I'm not sure if the if will be optimized, especially since you're checking inputs in one of the conditions.

PepperLola commented on fission/src/systems/scene/GizmoSceneObject.ts:
```diff
@@ -0,0 +1,194 @@
+        }
+
+        // creating enter key and escape key event listeners
+        if (InputSystem.isKeyPressed("Enter") && this._parentObject) {
+            // confirming placement of object
+            if (this._parentObject !== undefined) this._parentObject.EnablePhysics()
+            World.SceneRenderer.RemoveSceneObject(this.id)
+            return
+        } else if (InputSystem.isKeyPressed("Escape") && this._parentObject) {
+            // cancelling the creation of the mirabuf scene object
+            World.SceneRenderer.RemoveSceneObject(this._parentObject.id)
+            World.SceneRenderer.RemoveSceneObject(this.id)
+            return
+        }
```
Should `Dispose` be called somewhere in here instead of one of these calls? I'm genuinely asking since I don't know the answer, it very well could be no.

PepperLola commented on fission/src/systems/scene/GizmoSceneObject.ts:
```diff
@@ -0,0 +1,194 @@
+    public UpdateBodyPositionAndRotation(rn: RigidNodeReadOnly) {
+        if (!this._parentObject) return
+        World.PhysicsSystem.SetBodyPosition(
+            this._parentObject.mechanism.GetBodyByNodeId(rn.id)!,
```
Are we certain that the result of `GetBodyByNodeId` will be defined? I'd definitely hope so, but would it be worth instead checking that value and then calling `World.PhysicsSystem.SetBodyPosition` only if it's defined?

PepperLola commented on fission/src/systems/scene/GizmoSceneObject.ts:
```diff
@@ -0,0 +1,194 @@
+            ThreeMatrix4_JoltMat44(this._obj.matrix).GetTranslation()
+        )
+        World.PhysicsSystem.SetBodyRotation(
+            this._parentObject.mechanism.GetBodyByNodeId(rn.id)!,
```
Same here

PepperLola commented on fission/src/systems/scene/SceneRenderer.ts:
```diff
@@ -236,13 +231,29 @@ class SceneRenderer extends WorldSystem {
 
+    /** Registers gizmos that are attached to a parent mirabufsceneobject  */
+    public RegisterGizmoSceneObject(obj: GizmoSceneObject): number {
+        if (obj.HasParent()) this._gizmosOnMirabuf.set(obj.parentObjectId!, obj)
```
Will `obj.parentObjectId` necessarily be defined?

HunterBarclay replied:
`obj.HasParent()` essentially checks `obj.parent != undefined`.

HunterBarclay replied:
I suppose you could technically add an unregistered object to the gizmo

PepperLola commented on fission/src/systems/scene/SceneRenderer.ts:
```diff
@@ -236,13 +231,29 @@ class SceneRenderer extends WorldSystem {
         const obj = this._sceneObjects.get(id)
+
+        // If the object is a mirabuf object, remove the gizmo as well
+        if (obj instanceof MirabufSceneObject) {
+            const objGizmo = this._gizmosOnMirabuf.get(id)
+            if (this._gizmosOnMirabuf.delete(id)) objGizmo!.Dispose()
+        } else if (obj instanceof GizmoSceneObject && obj.HasParent()) {
+            this._gizmosOnMirabuf.delete(obj.parentObjectId!)
+        }
+
```
Will these necessarily be defined?

HunterBarclay replied:
Again, see the earlier comment.

PepperLola commented on fission/src/systems/scene/SceneRenderer.ts:
```diff
@@ -236,13 +231,29 @@ class SceneRenderer extends WorldSystem {
+        }
+
         if (this._sceneObjects.delete(id)) {
             obj!.Dispose()
```
Or `obj`?

HunterBarclay replied:
If delete succeeds, then `obj` would exist

PepperLola commented on fission/src/systems/scene/SceneRenderer.ts:
```diff
@@ -328,83 +339,11 @@ class SceneRenderer extends WorldSystem {
-            }
-        })
+    /** returns whether any gizmos are being currently dragged */
+    public IsAnyGizmoDragging(): boolean {
+        return Array.from(this._sceneObjects.values())
+            .filter(obj => obj instanceof GizmoSceneObject)
+            .some(obj => obj.gizmo.dragging)
     }
```
```suggestion
    public IsAnyGizmoDragging(): boolean {
        return Array.from(this._gizmosOnMirabuf.values())
            .some(gizmo => gizmo.dragging)
    }
```

I'm not sure if I got the syntax/names right but you can skip filtering since you're already keeping a map of active gizmos.

PepperLola commented on fission/src/ui/panels/configuring/assembly-config/interfaces/ConfigureGamepiecePickupInterface.tsx:
```diff
@@ -217,8 +219,8 @@ const ConfigureGamepiecePickupInterface: React.FC<ConfigPickupProps> = ({ select
                         )
-                        transformGizmo.mesh.position.setFromMatrixPosition(robotTransformation)
-                        transformGizmo.mesh.rotation.setFromRotationMatrix(robotTransformation)
+                        transformGizmo.obj.position.setFromMatrixPosition(robotTransformation)
+                        transformGizmo.obj.rotation.setFromRotationMatrix(robotTransformation)
```
```suggestion
                        transformGizmo.UpdateGizmoObjectPositionAndRotation(robotTransformation)
```

PepperLola commented on fission/src/ui/panels/configuring/assembly-config/interfaces/ConfigureGamepiecePickupInterface.tsx:
```diff
@@ -107,7 +112,7 @@ const ConfigureGamepiecePickupInterface: React.FC<ConfigPickupProps> = ({ select
         }
 
-        transformGizmo.mesh.scale.set(zoneSize, zoneSize, zoneSize)
+        transformGizmo.obj.scale.set(zoneSize, zoneSize, zoneSize)
```
I don't like that we're manipulating internal gizmo objects from outside of it. Maybe there should be a method to set the scale of the gizmo from out here? That way if the implementation of the gizmo changes, we would only need to make sure all the methods inside the class still work as expected, rather than needing to track down all the places where we made assumptions like this about the internal implementation of the class.

PepperLola commented on fission/src/ui/panels/configuring/assembly-config/interfaces/ConfigureGamepiecePickupInterface.tsx:
```diff
@@ -117,16 +122,16 @@ const ConfigureGamepiecePickupInterface: React.FC<ConfigPickupProps> = ({ select
-        ;(gizmo.mesh.material as THREE.Material).depthTest = false
-        gizmo.AddMeshToScene()
-        gizmo.CreateGizmo("translate", 1.5)
+        ;((gizmo.obj as THREE.Mesh).material as THREE.Material).depthTest = false
```
Same thing here. We could avoid the whole ESLint vs Prettier disagreement about lines starting with a semicolon as well as the double cast and generally ugly line by creating a function to set the `depthTest` field of the material. Also, is this something that needs to happen every time the gizmo is created? If so, it would be worth putting it inside of some function in the gizmo object so that we don't need to remember to call it every time.

PepperLola commented on fission/src/ui/panels/configuring/assembly-config/interfaces/ConfigureShotTrajectoryInterface.tsx:
```diff
@@ -108,21 +108,17 @@ const ConfigureShotTrajectoryInterface: React.FC<ConfigEjectorProps> = ({ select
-        gizmo.CreateGizmo("translate", 1.5)
-        gizmo.CreateGizmo("rotate", 2.0)
+        if (!mesh) return
+        ;(mesh.material as THREE.Material).depthTest = false
```
```suggestion
        const material: THREE.Material = mesh.material
        material.depthTest = false
```

I like the look of this better but it doesn't really matter.

PepperLola commented on fission/src/ui/panels/configuring/assembly-config/interfaces/ConfigureShotTrajectoryInterface.tsx:
```diff
@@ -208,12 +208,12 @@ const ConfigureShotTrajectoryInterface: React.FC<ConfigEjectorProps> = ({ select
                         )
-                        transformGizmo.mesh.position.setFromMatrixPosition(robotTransformation)
-                        transformGizmo.mesh.rotation.setFromRotationMatrix(robotTransformation)
+                        ejectorMesh.position.setFromMatrixPosition(robotTransformation)
+                        ejectorMesh.rotation.setFromRotationMatrix(robotTransformation)
```
There's a lot of repetition of these two functions. Would it be worth creating a function to do that for you? There's already one in the gizmo class but that only works if the mesh is stored in the gizmo.

PepperLola commented on fission/src/ui/panels/configuring/assembly-config/interfaces/scoring/ZoneConfigInterface.tsx:
```diff
@@ -71,7 +71,7 @@ function save(
     const rotation = new THREE.Quaternion(0, 0, 0, 1)
     const scale = new THREE.Vector3(1, 1, 1)
-    gizmo.mesh.matrixWorld.decompose(translation, rotation, scale)
+    gizmo.obj.matrixWorld.decompose(translation, rotation, scale)
```
I still don't like that we're doing this from outside the gizmo.

HunterBarclay replied:
At the end of the day, whatever uses the gizmo needs to have the freedom to do whatever it needs with the transform data it provides. I don't think there will ever be a way around this purely.

PepperLola commented on fission/src/ui/panels/configuring/assembly-config/interfaces/scoring/ZoneConfigInterface.tsx:
```diff
@@ -170,13 +170,13 @@ const ZoneConfigInterface: React.FC<ZoneConfigProps> = ({ selectedField, selecte
-        ;(gizmo.mesh.material as THREE.Material).depthTest = false
-        gizmo.AddMeshToScene()
-        gizmo.CreateGizmo("translate", 1.5)
+        ;((gizmo.obj as THREE.Mesh).material as THREE.Material).depthTest = false
```
Same as above

PepperLola commented on fission/src/ui/panels/configuring/assembly-config/interfaces/scoring/ZoneConfigInterface.tsx:
```diff
@@ -190,14 +190,14 @@ const ZoneConfigInterface: React.FC<ZoneConfigProps> = ({ selectedField, selecte
-        gizmo.mesh.position.set(props.translation.x, props.translation.y, props.translation.z)
-        gizmo.mesh.rotation.setFromQuaternion(props.rotation)
-        gizmo.mesh.scale.set(props.scale.x, props.scale.y, props.scale.z)
+        gizmo.obj.position.set(props.translation.x, props.translation.y, props.translation.z)
+        gizmo.obj.rotation.setFromQuaternion(props.rotation)
+        gizmo.obj.scale.set(props.scale.x, props.scale.y, props.scale.z)
```
I don't think we should be directly mutating properties of the gizmo from outside.

PepperLola reviewed (changes requested):
For any of the non-null assertions, if you're absolutely certain that they're valid and will never cause issues, please respond with your justification and then resolve the comment. There are also some stylistic suggestions that I made that you don't need to accept if you or other people prefer how they are now.

HunterBarclay committed:
Initial gizmo component progress

HunterBarclay committed:
fix!: Locate the issue with using the transform gizmo component.

WIP

HunterBarclay committed:
refactor: Migrate use of gizmo in config UI to new gizmo component.

WIP: Migrating scheme selection to a component atm.

HunterBarclay committed:
refactor: Move input scheme selection into new component.

WIP: Making a new panel combining input scheme selection and assembly
transformation into one.

HunterBarclay committed:
Merge branch 'dev' into dhruv/1798/refactoring-gizmos

HunterBarclay committed:
feat: Add initial config menu that includes gizmo controls.

This feature also adds additional upgrades and refactoring to how
pausing is handled. It adds the side-effect of when an assembly fails to
be spawned in, the simulation will remain paused for approx. 500 ms.

Also includes global UI controls that can be accessed from anywhere.

HunterBarclay committed:
feat: Add assembly moving to configuration, along with move panel.

HunterBarclay committed:
fix: Gizmo centers on selected assembly.

HunterBarclay committed:
fix: Added new initial config panel to import local.

Also added in a couple cleanups here and there.

HunterBarclay committed:
fix: Move assemblies into the center of the field.

Also finally moved the ground up as well.

HunterBarclay committed:
Format plus some syntax errors

HunterBarclay committed:
fix: Extra semi-colons from prettier

HunterBarclay committed:
refactor: Use optional instead of union type.

Co-authored-by: PepperLola <PepperLola@users.noreply.github.com>

HunterBarclay committed:
Remove if statement

Co-authored-by: PepperLola <PepperLola@users.noreply.github.com>

HunterBarclay committed:
fix: Make changes suggested by @PepperLola .

Also included some snapping controls.

HunterBarclay committed:
Format

HunterBarclay committed:
fix: Readd delete controls to buttons

WIP!: Currently not working for removing input schemes from deleted
robots.

HunterBarclay committed:
fix: Delete and confirm work as expected.

HunterBarclay committed:
feat(gizmos): Add escape and enter keys for gizmo

The TransformGizmoContro now contains callbacks for when it "receives"
the signal to cancel or accept it's position via the escape and enter
keys respectively.

HunterBarclay committed:
chore: Format

HunterBarclay committed:
chore: Python type linting
